pico.py

```diff
@@ -6,43 +6,126 @@
 HOST='127.0.0.1'
 PORT=5000
 
+def is_connection_closed(conn) -> bool:
+    try:
+        # this will try to read bytes without blocking and also without removing them from buffer (peek only)
+        data = conn.recv(4, socket.MSG_DONTWAIT | socket.MSG_PEEK)
+        if len(data) == 0:
+            return True
+    except BlockingIOError:
+        return False  # socket is open and reading from it would block
+    except ConnectionResetError:
+        return True  # socket was closed for some other reason
+    except Exception as e:
+        return False
+    return False
+
+def is_valid_command(command):
+	command_length = len(bin(command)[2:])
+
+	if command_length < 24:
+		print(f"Invalid command: {bin(command)}")
+		return False
+	else:
+		return True
+	
+
+def get_command_header(command):
+	# move all bits except the first 4 to the right
+	# only the first 4 bits will be returned
+
+	if not is_valid_command(command):
+		return 0
+
+	return command >> len(bin(command)[2:]) - 4
+
+def get_command_data(command):
+	if not is_valid_command(command):
+		return 0
+	
+	header_and_data = command >> len(bin(command)[2:]) - 24
+	return header_and_data & int('0' * 4 + '1' * 20, 2)
+
 def signal_handler(signal, frame):
 	global connection
-	with connection:
-		print("\nSending closing signal...")
-		connection.sendall(b'\x00\x00\x00\x00')
-		sys.exit(0)
+	try:
+		with connection:
+			print("\nSending closing signal...")
+			connection.sendall(b'\x00\x00\x00\x00')
+	except:
+		print("\nNo client was connected")
+	sys.exit(0)
```

Comment:
When you press control c tell the connected client. If it's not there just exit.